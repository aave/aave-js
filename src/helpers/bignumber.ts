import BigNumber from 'bignumber.js';

export type BigNumberValue = string | number | BigNumber;

export const BigNumberZD = BigNumber.clone({
  DECIMAL_PLACES: 0,
  ROUNDING_MODE: BigNumber.ROUND_DOWN,
});

export function valueToBigNumber(amount: BigNumberValue): BigNumber {
  return new BigNumber(amount);
}
export function valueToZDBigNumber(amount: BigNumberValue): BigNumber {
  return new BigNumberZD(amount);
}
