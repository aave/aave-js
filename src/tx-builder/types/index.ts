import { providers, BigNumber, BytesLike, PopulatedTransaction } from 'ethers';

export type tEthereumAddress = string;
export type tStringCurrencyUnits = string; // ex 2.5 eth
export type tStringDecimalUnits = string; // ex 2500000000000000000
export type ENS = string; // something.eth

/** InterestRate options */
export enum InterestRate {
  None = 'None',
  Stable = 'Stable',
  Variable = 'Variable',
}

export enum Market {
  Proto = 'proto',
  AMM = 'amm',
  Arc = 'arc',
}

export enum Network {
  mainnet = 'mainnet',
  ropsten = 'ropsten',
  kovan = 'kovan',
  polygon = 'polygon',
  fork = 'fork',
  mumbai = 'mumbai',
  polygon_fork = 'polygon_fork',
}

export enum ChainId {
  mainnet = 1,
  ropsten = 3,
  kovan = 42,
  polygon = 137,
  fork = 1337,
  mumbai = 80001,
  polygon_fork = 1338,
}

export enum eEthereumTxType {
  ERC20_APPROVAL = 'ERC20_APPROVAL',
  DLP_ACTION = 'DLP_ACTION',
  GOVERNANCE_ACTION = 'GOVERNANCE_ACTION',
  GOV_DELEGATION_ACTION = 'GOV_DELEGATION_ACTION',
  STAKE_ACTION = 'STAKE_ACTION',
  MIGRATION_LEND_AAVE = 'MIGRATION_LEND_AAVE',
  FAUCET_MINT = 'FAUCET_MINT',
  REWARD_ACTION = 'REWARD_ACTION',
}

export enum ProtocolAction {
  default = 'default',
  withdraw = 'withdraw',
  deposit = 'deposit',
  liquidationCall = 'liquidationCall',
  liquidationFlash = 'liquidationFlash',
  repay = 'repay',
  swapCollateral = 'swapCollateral',
  repayCollateral = 'repayCollateral',
  withdrawETH = 'withdrawETH',
  borrowETH = 'borrwoETH',
}

export enum GovernanceVote {
  Abstain = 0,
  Yes = 1,
  No = 2,
}

export enum Stake {
  Aave = 'Aave',
  Balancer = 'Balancer',
}

export type GasRecommendationType = {
  [action: string]: {
    limit: string;
    recommended: string;
  };
};

export type GeneratedTx = {
  tx: transactionType;
  gas: {
    price: string;
    limit: string;
  };
};

export type transactionType = {
  value?: string;
  from?: string;
  to?: string;
  nonce?: number;
  gasLimit?: BigNumber;
  gasPrice?: BigNumber;
  data?: string;
  chainId?: number;
};

export type AddressModel = {
  ADDRESS_PROVIDER_ADDRESS: tEthereumAddress;
  LENDINGPOOL_ADDRESS: tEthereumAddress;
  LENDINGPOOL_CORE_ADDRESS: tEthereumAddress;
  SYNTHETIX_PROXY_ADDRESS: tEthereumAddress;
  GOVERNANCE_PROTO_CONTRACT: tEthereumAddress;
  LEND_TO_AAVE_MIGRATOR: tEthereumAddress;
  WETH_GATEWAY: tEthereumAddress;
  FAUCET: tEthereumAddress;
  SWAP_COLLATERAL_ADAPTER: tEthereumAddress;
  REPAY_WITH_COLLATERAL_ADAPTER: tEthereumAddress;
  AAVE_GOVERNANCE_V2: tEthereumAddress;
  AAVE_GOVERNANCE_V2_EXECUTOR_SHORT: tEthereumAddress;
  AAVE_GOVERNANCE_V2_EXECUTOR_LONG: tEthereumAddress;
  AAVE_GOVERNANCE_V2_HELPER: tEthereumAddress;
  FLASHLIQUIDATION: tEthereumAddress;
  INCENTIVES_CONTROLLER: tEthereumAddress;
  INCENTIVES_CONTROLLER_REWARD_TOKEN: tEthereumAddress;
};

export type tCommonContractAddressBetweenMarkets = Pick<
  AddressModel,
  | 'SYNTHETIX_PROXY_ADDRESS'
  | 'GOVERNANCE_PROTO_CONTRACT'
  | 'LEND_TO_AAVE_MIGRATOR'
  | 'FAUCET'
  | 'SWAP_COLLATERAL_ADAPTER'
  | 'REPAY_WITH_COLLATERAL_ADAPTER'
  | 'FLASHLIQUIDATION'
  | 'INCENTIVES_CONTROLLER'
  | 'INCENTIVES_CONTROLLER_REWARD_TOKEN'
>;

export type tDistinctContractAddressBetweenMarkets = Pick<
  AddressModel,
  | 'ADDRESS_PROVIDER_ADDRESS'
  | 'LENDINGPOOL_ADDRESS'
  | 'LENDINGPOOL_CORE_ADDRESS'
>;

export type tDistinctContractAddressBetweenMarketsV2 = Pick<
  AddressModel,
  'LENDINGPOOL_ADDRESS' | 'WETH_GATEWAY'
>;

export type tDistinctGovernanceV2Addresses = Pick<
  AddressModel,
  | 'AAVE_GOVERNANCE_V2'
  | 'AAVE_GOVERNANCE_V2_EXECUTOR_SHORT'
  | 'AAVE_GOVERNANCE_V2_EXECUTOR_LONG'
  | 'AAVE_GOVERNANCE_V2_HELPER'
>;

export type tdistinctStakingAddressesBetweenTokens = {
  TOKEN_STAKING_ADDRESS: tEthereumAddress;
  STAKING_REWARD_TOKEN_ADDRESS: tEthereumAddress;
  STAKING_HELPER_ADDRESS: tEthereumAddress;
  canUsePermit: boolean;
};

export type ContractAddresses = {
  [contractName: string]: tEthereumAddress;
};

export type Configuration = {
  network: Network;
  provider:
    | providers.JsonRpcProvider
    | providers.BaseProvider
    | providers.Web3Provider;
};

export type EthereumTransactionTypeExtended = {
  txType: eEthereumTxType;
  tx: () => Promise<transactionType>;
  gas: GasResponse;
};

export type TransactionGenerationMethod = {
  rawTxMethod: () => Promise<PopulatedTransaction>;
  from: tEthereumAddress;
  value?: string;
  gasSurplus?: number;
  action?: ProtocolAction;
};

export type TransactionGasGenerationMethod = {
  txCallback: () => Promise<transactionType>;
  action?: ProtocolAction;
};

export type GasType = {
  gasLimit: string | undefined;
  gasPrice: string;
};
export type GasResponse = (force?: boolean) => Promise<GasType | null>;

export type TokenMetadataType = {
  name: string;
  symbol: string;
  decimals: number;
  address: string;
};

export type DefaultProviderKeys = {
  etherscan?: string;
  infura?: string;
  alchemy?: string;
};

export type GovernanceConfigType = {
  [network: string]: tDistinctGovernanceV2Addresses;
};
export type StakingConfigType = {
  [sToken: string]: {
    [network: string]: tdistinctStakingAddressesBetweenTokens;
  };
};

export type CommonConfigType = {
  [network: string]: tCommonContractAddressBetweenMarkets;
};

export type LendingPoolConfigType = {
  [pool: string]: {
    [network: string]: tDistinctContractAddressBetweenMarketsV2;
  };
};

export type EnabledNetworksType = {
  staking: {
    [sToken: string]: Network[];
  };
  lendingPool: {
    [market: string]: Network[];
  };
  governance: Network[];
  wethGateway: { [market: string]: Network[] };
  faucet: Network[];
  liquiditySwapAdapter: Network[];
  repayWithCollateralAdapter: Network[];
  aaveGovernanceV2: Network[];
  ltaMigrator: Network[];
  incentivesController: Network[];
};

export type PermitSignature = {
  amount: tStringCurrencyUnits;
  deadline: string;
  v: number;
  r: BytesLike;
  s: BytesLike;
};

export type FlashLoanParams = {
  assetToSwapToList: tEthereumAddress[]; // List of the addresses of the reserve to be swapped to and deposited
  minAmountsToReceive: string[]; // List of min amounts to be received from the swap
  swapAllBalance: boolean[]; // Flag indicating if all the user balance should be swapped
  permitAmount: string[]; // List of amounts for the permit signature
  deadline: string[]; // List of deadlines for the permit signature
  v: number[]; // List of v param for the permit signature
  r: BytesLike[]; // List of r param for the permit signature
  s: BytesLike[]; // List of s param for the permit signature
};
