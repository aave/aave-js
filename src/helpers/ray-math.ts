import BigNumber from 'bignumber.js';
import { BigNumberValue, valueToZDBigNumber } from './bignumber';

export const WAD = valueToZDBigNumber(10).pow(18);
export const HALF_WAD = WAD.dividedBy(2);

export const RAY = valueToZDBigNumber(10).pow(27);
export const HALF_RAY = RAY.dividedBy(2);

export const WAD_RAY_RATIO = valueToZDBigNumber(10).pow(9);

export function wadMul(a: BigNumberValue, b: BigNumberValue): BigNumber {
  return HALF_WAD.plus(valueToZDBigNumber(a).multipliedBy(b)).div(WAD);
}

export function wadDiv(a: BigNumberValue, b: BigNumberValue): BigNumber {
  const halfB = valueToZDBigNumber(b).div(2);

  return halfB.plus(valueToZDBigNumber(a).multipliedBy(WAD)).div(b);
}

export function rayMul(a: BigNumberValue, b: BigNumberValue): BigNumber {
  return HALF_RAY.plus(valueToZDBigNumber(a).multipliedBy(b)).div(RAY);
}

export function rayDiv(a: BigNumberValue, b: BigNumberValue): BigNumber {
  const halfB = valueToZDBigNumber(b).div(2);

  return halfB.plus(valueToZDBigNumber(a).multipliedBy(RAY)).div(b);
}

export function rayToWad(a: BigNumberValue): BigNumber {
  const halfRatio = valueToZDBigNumber(WAD_RAY_RATIO).div(2);

  return halfRatio.plus(a).div(WAD_RAY_RATIO);
}

export function wadToRay(a: BigNumberValue): BigNumber {
  return valueToZDBigNumber(a).multipliedBy(WAD_RAY_RATIO).decimalPlaces(0);
}

export function rayPow(a: BigNumberValue, p: BigNumberValue): BigNumber {
  let x = valueToZDBigNumber(a);
  let n = valueToZDBigNumber(p);
  let z = !n.modulo(2).eq(0) ? x : valueToZDBigNumber(RAY);

  for (n = n.div(2); !n.eq(0); n = n.div(2)) {
    x = rayMul(x, x);

    if (!n.modulo(2).eq(0)) {
      z = rayMul(z, x);
    }
  }
  return z;
}

/**
 * RayPow is slow and gas intensive therefore in v2 we switched to binomial approximation on the contract level.
 * While the results ar not exact to the last decimal, they are close enough.
 */
export function binomialApproximatedRayPow(
  a: BigNumberValue,
  p: BigNumberValue
): BigNumber {
  const base = valueToZDBigNumber(a);
  const exp = valueToZDBigNumber(p);
  if (exp.eq(0)) return RAY;
  const expMinusOne = exp.minus(1);
  const expMinusTwo = exp.gt(2) ? exp.minus(2) : 0;

  const basePowerTwo = rayMul(base, base);
  const basePowerThree = rayMul(basePowerTwo, base);

  const firstTerm = exp.multipliedBy(base);
  const secondTerm = exp
    .multipliedBy(expMinusOne)
    .multipliedBy(basePowerTwo)
    .div(2);
  const thirdTerm = exp
    .multipliedBy(expMinusOne)
    .multipliedBy(expMinusTwo)
    .multipliedBy(basePowerThree)
    .div(6);

  return RAY.plus(firstTerm).plus(secondTerm).plus(thirdTerm);
}

export function rayToDecimal(a: BigNumberValue): BigNumber {
  return valueToZDBigNumber(a).dividedBy(RAY);
}
