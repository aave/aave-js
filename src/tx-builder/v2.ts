import { providers } from 'ethers';
import { Network, DefaultProviderKeys, TxBuilderConfig } from './types';
import TxBuilderInterface from './interfaces/TxBuilder';
import LendingPoolInterface from './interfaces/v2/LendingPool';
import LendingPool from './services/v2/LendingPool';
import BaseTxBuilder from './txBuilder';
import WETHGatewayInterface from './interfaces/WETHGateway';
import WETHGatewayService from './services/WETHGateway';
import BaseDebtTokenInterface from './interfaces/BaseDebtToken';
import BaseDebtToken from './services/BaseDebtToken';
import LiquiditySwapAdapterService from './services/LiquiditySwapAdapterParaswap';
import LiquiditySwapAdapterInterface from './interfaces/LiquiditySwapAdapterParaswap';
import RepayWithCollateralAdapterService from './services/RepayWithCollateralAdapter';
import RepayWithCollateralAdapterInterface from './interfaces/RepayWithCollateralAdapter';
import AaveGovernanceV2Interface from './interfaces/v2/AaveGovernanceV2';
import GovernanceDelegationTokenInterface from './interfaces/v2/GovernanceDelegationToken';
import AaveGovernanceV2Service from './services/v2/AaveGovernanceV2';
import GovernanceDelegationTokenService from './services/v2/GovernanceDelegationTokenService';

export default class TxBuilder
  extends BaseTxBuilder
  implements TxBuilderInterface {
  readonly lendingPools: {
    [market: string]: LendingPoolInterface;
  };

  readonly wethGateways: {
    [market: string]: WETHGatewayInterface;
  };

  readonly swapCollateralAdapters: {
    [market: string]: LiquiditySwapAdapterInterface;
  };

  readonly repayWithCollateralAdapters: {
    [market: string]: RepayWithCollateralAdapterInterface;
  };

  readonly baseDebtTokenService: BaseDebtTokenInterface;

  public aaveGovernanceV2Service: AaveGovernanceV2Interface;

  public governanceDelegationTokenService: GovernanceDelegationTokenInterface;

  constructor(
    network: Network = Network.mainnet,
    injectedProvider?: providers.Provider | string | undefined,
    defaultProviderKeys?: DefaultProviderKeys,
    config?: TxBuilderConfig
  ) {
    super(network, injectedProvider, defaultProviderKeys, config);

    this.wethGateways = {};
    this.lendingPools = {};
    this.swapCollateralAdapters = {};
    this.repayWithCollateralAdapters = {};
    this.baseDebtTokenService = new BaseDebtToken(
      this.configuration,
      this.erc20Service
    );

    this.aaveGovernanceV2Service = new AaveGovernanceV2Service(
      this.configuration,
      this.txBuilderConfig.governance?.[network]
    );

    this.governanceDelegationTokenService = new GovernanceDelegationTokenService(
      this.configuration
    );
  }

  public getRepayWithCollateralAdapter = (
    market: string
  ): RepayWithCollateralAdapterInterface => {
    const { network } = this.configuration;

    if (!this.repayWithCollateralAdapters[market]) {
      this.repayWithCollateralAdapters[
        market
      ] = new RepayWithCollateralAdapterService(
        this.configuration,
        this.txBuilderConfig.lendingPool?.[network]?.[market]
      );
    }

    return this.repayWithCollateralAdapters[market];
  };

  public getSwapCollateralAdapter = (
    market: string
  ): LiquiditySwapAdapterInterface => {
    const { network } = this.configuration;

    if (!this.swapCollateralAdapters[market]) {
      this.swapCollateralAdapters[market] = new LiquiditySwapAdapterService(
        this.configuration,
        this.txBuilderConfig.lendingPool?.[network]?.[market]
      );
    }

    return this.swapCollateralAdapters[market];
  };

  public getWethGateway = (market: string): WETHGatewayInterface => {
    const { network } = this.configuration;
    if (!this.wethGateways[market]) {
      this.wethGateways[market] = new WETHGatewayService(
        this.configuration,
        this.baseDebtTokenService,
        this.erc20Service,
        this.txBuilderConfig.lendingPool?.[network]?.[market]
      );
    }

    return this.wethGateways[market];
  };

  public getLendingPool = (market: string): LendingPoolInterface => {
    const { network } = this.configuration;
    if (!this.lendingPools[market]) {
      this.lendingPools[market] = new LendingPool(
        this.configuration,
        this.erc20Service,
        this.synthetixService,
        this.getWethGateway(market),
        this.getSwapCollateralAdapter(market),
        this.getRepayWithCollateralAdapter(market),
        market,
        this.txBuilderConfig.lendingPool?.[network]?.[market]
      );
    }

    return this.lendingPools[market];
  };
}
