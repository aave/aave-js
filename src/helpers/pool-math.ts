import BigNumber from 'bignumber.js';

import {
  BigNumberValue,
  valueToBigNumber,
  valueToZDBigNumber,
  pow10,
} from './bignumber';
import * as RayMath from './ray-math';
import { SECONDS_PER_YEAR } from './constants';

export const LTV_PRECISION = 4;

export function calculateCompoundedInterest(
  rate: BigNumberValue,
  currentTimestamp: number,
  lastUpdateTimestamp: number
): BigNumber {
  const timeDelta = valueToZDBigNumber(currentTimestamp - lastUpdateTimestamp);
  const ratePerSecond = valueToZDBigNumber(rate).dividedBy(SECONDS_PER_YEAR);
  return RayMath.binomialApproximatedRayPow(ratePerSecond, timeDelta);
}

export function getCompoundedBalance(
  _principalBalance: BigNumberValue,
  _reserveIndex: BigNumberValue,
  _reserveRate: BigNumberValue,
  _lastUpdateTimestamp: number,
  currentTimestamp: number
): BigNumber {
  const principalBalance = valueToZDBigNumber(_principalBalance);
  if (principalBalance.eq('0')) {
    return principalBalance;
  }

  const compoundedInterest = calculateCompoundedInterest(
    _reserveRate,
    currentTimestamp,
    _lastUpdateTimestamp
  );
  const cumulatedInterest = RayMath.rayMul(compoundedInterest, _reserveIndex);
  const principalBalanceRay = RayMath.wadToRay(principalBalance);

  return RayMath.rayToWad(
    RayMath.rayMul(principalBalanceRay, cumulatedInterest)
  );
}

export const calculateLinearInterest = (
  rate: BigNumberValue,
  currentTimestamp: number,
  lastUpdateTimestamp: number
) => {
  const timeDelta = RayMath.wadToRay(
    valueToZDBigNumber(currentTimestamp - lastUpdateTimestamp)
  );
  const timeDeltaInSeconds = RayMath.rayDiv(
    timeDelta,
    RayMath.wadToRay(SECONDS_PER_YEAR)
  );
  return RayMath.rayMul(rate, timeDeltaInSeconds).plus(RayMath.RAY);
};

export function getReserveNormalizedIncome(
  rate: BigNumberValue,
  index: BigNumberValue,
  lastUpdateTimestamp: number,
  currentTimestamp: number
): BigNumber {
  if (valueToZDBigNumber(rate).eq('0')) {
    return valueToZDBigNumber(index);
  }

  const cumulatedInterest = calculateLinearInterest(
    rate,
    currentTimestamp,
    lastUpdateTimestamp
  );

  return RayMath.rayMul(cumulatedInterest, index);
}

export function getLinearBalance(
  balance: BigNumberValue,
  index: BigNumberValue,
  rate: BigNumberValue,
  lastUpdateTimestamp: number,
  currentTimestamp: number
) {
  return RayMath.rayToWad(
    RayMath.rayMul(
      RayMath.wadToRay(balance),
      getReserveNormalizedIncome(
        rate,
        index,
        lastUpdateTimestamp,
        currentTimestamp
      )
    )
  );
}

export function getCompoundedStableBalance(
  _principalBalance: BigNumberValue,
  _userStableRate: BigNumberValue,
  _lastUpdateTimestamp: number,
  currentTimestamp: number
): BigNumber {
  const principalBalance = valueToZDBigNumber(_principalBalance);
  if (principalBalance.eq('0')) {
    return principalBalance;
  }

  const cumulatedInterest = calculateCompoundedInterest(
    _userStableRate,
    currentTimestamp,
    _lastUpdateTimestamp
  );
  const principalBalanceRay = RayMath.wadToRay(principalBalance);

  return RayMath.rayToWad(
    RayMath.rayMul(principalBalanceRay, cumulatedInterest)
  );
}

export function calculateHealthFactorFromBalances(
  collateralBalanceETH: BigNumberValue,
  borrowBalanceETH: BigNumberValue,
  currentLiquidationThreshold: BigNumberValue
): BigNumber {
  if (valueToBigNumber(borrowBalanceETH).eq(0)) {
    return valueToBigNumber('-1'); // invalid number
  }
  return valueToBigNumber(collateralBalanceETH)
    .multipliedBy(currentLiquidationThreshold)
    .dividedBy(pow10(LTV_PRECISION))
    .div(borrowBalanceETH);
}

export function calculateHealthFactorFromBalancesBigUnits(
  collateralBalanceETH: BigNumberValue,
  borrowBalanceETH: BigNumberValue,
  currentLiquidationThreshold: BigNumberValue
): BigNumber {
  return calculateHealthFactorFromBalances(
    collateralBalanceETH,
    borrowBalanceETH,
    new BigNumber(currentLiquidationThreshold)
      .multipliedBy(pow10(LTV_PRECISION))
      .decimalPlaces(0, BigNumber.ROUND_DOWN)
  );
}

export function calculateAvailableBorrowsETH(
  collateralBalanceETH: BigNumberValue,
  borrowBalanceETH: BigNumberValue,
  currentLtv: BigNumberValue
): BigNumber {
  if (valueToZDBigNumber(currentLtv).eq(0)) {
    return valueToZDBigNumber('0');
  }
  const availableBorrowsETH = valueToZDBigNumber(collateralBalanceETH)
    .multipliedBy(currentLtv)
    .dividedBy(pow10(LTV_PRECISION))
    .minus(borrowBalanceETH);
  return availableBorrowsETH.gt('0')
    ? availableBorrowsETH
    : valueToZDBigNumber('0');
}

export function calculateAverageRate(
  index0: string,
  index1: string,
  timestamp0: number,
  timestamp1: number
): string {
  return valueToBigNumber(index1)
    .dividedBy(index0)
    .minus('1')
    .dividedBy(timestamp1 - timestamp0)
    .multipliedBy(SECONDS_PER_YEAR)
    .toString();
}
