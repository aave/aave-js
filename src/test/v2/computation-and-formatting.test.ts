import { ReserveData, UserReserveData } from '../../v2/types';
import {
  formatReserves,
  formatUserSummaryData,
} from '../../v2/computations-and-formatting';
import BigNumber from 'bignumber.js';
import { getCompoundedBalance } from '../../helpers/pool-math';

const mockReserve: ReserveData = {
  underlyingAsset: '0xff795577d9ac8bd7d90ee22b6c1703490b6512fd',
  name: '',
  symbol: 'DAI',
  decimals: 18,
  baseLTVasCollateral: '7500',
  reserveLiquidationThreshold: '8000',
  reserveLiquidationBonus: '10500',
  reserveFactor: '1000',
  usageAsCollateralEnabled: true,
  borrowingEnabled: true,
  stableBorrowRateEnabled: true,
  isActive: true,
  isFrozen: false,
  liquidityIndex: '1000164447379610590574518134',
  variableBorrowIndex: '1000232854433711209646283880',
  liquidityRate: '26776200735312093055313462',
  variableBorrowRate: '38568743388028395681971229',
  stableBorrowRate: '109284371694014197840985614',
  lastUpdateTimestamp: 1606992400,
  aTokenAddress: '0xdCf0aF9e59C002FA3AA091a46196b37530FD48a8',
  stableDebtTokenAddress: '0x3B91257Fe5CA63b4114ac41A0d467D25E2F747F3',
  variableDebtTokenAddress: '0xEAbBDBe7aaD7d5A278da40967E62C8c8Fe5fAec8',
  // interestRateStrategyAddress: '0x1c4c4dD7F19738Fd7C21Fa7CbF9667710ff3Ba4c',
  availableLiquidity: '43133641118657852003256',
  totalPrincipalStableDebt: '1000000000000000000',
  averageStableRate: '109284236984257451326752610',
  stableDebtLastUpdateTimestamp: 1606992400,
  totalScaledVariableDebt: '145496831599325217573288',
  // priceInEth: '1634050000000000',
  variableRateSlope1: '40000000000000000000000000',
  variableRateSlope2: '750000000000000000000000000',
  stableRateSlope1: '20000000000000000000000000',
  stableRateSlope2: '750000000000000000000000000',
  id:
    '0xff795577d9ac8bd7d90ee22b6c1703490b6512fd0x88757f2f99175387ab4c6a4b3067c77a695b0349',
  price: { priceInEth: '1634050000000000' },
  // fake data
  optimalUtilisationRate: '109284236984257451326752610',
  baseVariableBorrowRate: '109284236984257451326752610',
};

const mockUserReserve: UserReserveData = {
  // underlyingAsset: '0xFf795577d9AC8bD7D90Ee22b6C1703490b6512FD',
  scaledATokenBalance: '0',
  usageAsCollateralEnabledOnUser: false,
  stableBorrowRate: '0',
  scaledVariableDebt: '137602026075945229933190',
  principalStableDebt: '0',
  stableBorrowLastUpdateTimestamp: 0,
  variableBorrowIndex: '0',
  //id:0cd96fb5ee9616f64d892644f53f35be4f90xff795577d9ac8bd7d90ee22b6c1703490b6512fd0x88757f2f99175387ab4c6a4b3067c77a695b0349',
  reserve: {
    id:
      '0xff795577d9ac8bd7d90ee22b6c1703490b6512fd0x88757f2f99175387ab4c6a4b3067c77a695b0349',
    underlyingAsset: '0xff795577d9ac8bd7d90ee22b6c1703490b6512fd',
    name: '',
    symbol: 'DAI',
    decimals: 18,
    liquidityRate: '27070210296933117164358651',
    reserveLiquidationBonus: '10500',
    lastUpdateTimestamp: 1607000068,
  },
};

describe('computations and formattings', () => {
  describe('formatUserSummaryData', () => {
    const formattedMockReserve = formatUserSummaryData(
      [mockReserve],
      [mockUserReserve],
      '0cd96fb5ee9616f64d892644f53f35be4f90xff795577d9ac8bd7d90ee22b6c1703490b6512fd0x88757f2f99175387ab4c6a4b3067c77a695b0349',
      '598881655557838',
      mockUserReserve.reserve.lastUpdateTimestamp + 2000
    );
    expect(
      formattedMockReserve.reservesData[0].reserve.reserveLiquidationBonus
    ).toBe('0.05');
    expect(formattedMockReserve).toMatchSnapshot();
  });

  describe('formatReserves', () => {
    it('should return plausible results', () => {
      const formattedMockReserve = formatReserves(
        [mockReserve],
        mockReserve.lastUpdateTimestamp + 2000
      )[0];
      expect(formattedMockReserve.baseLTVasCollateral).toBe('0.75');
      expect(formattedMockReserve).toMatchSnapshot();
    });

    it('should allow omitting timestamp', () => {
      const formattedMockReserve = formatReserves([mockReserve])[0];
      expect(formattedMockReserve.baseLTVasCollateral).toBe('0.75');
      expect(formattedMockReserve).toMatchSnapshot();
    });

    /**
     * Whenever we add a new asset there#s a chance that an asset has no paramsHistory from 30days ago
     * We should not throw if that's the case, but just ignore it
     */
    it("should not error When 30dago reserves doesn't contain paramsHistory", () => {
      formatReserves([mockReserve], mockReserve.lastUpdateTimestamp + 2000, [
        mockReserve as any,
      ]);
      formatReserves([mockReserve], mockReserve.lastUpdateTimestamp + 2000, [
        { ...mockReserve, paramsHistory: [] } as any,
      ]);
      // would be wrong if any of the above threw
      expect(true).toBe(true);
    });

    it('should return proper values for new reserves', () => {
      const newReserve: Partial<ReserveData> = {
        availableLiquidity: '0',
        totalPrincipalStableDebt: '0',
        totalScaledVariableDebt: '0',
      };
      const formattedMockReserve = formatReserves(
        [
          {
            ...mockReserve,
            ...newReserve,
          },
        ],
        mockReserve.lastUpdateTimestamp
      )[0];
      expect(formattedMockReserve.utilizationRate).toBe('0');
    });

    it('should increase over time', () => {
      /**
       * tests against a regression which switched two dates
       */
      const first = formatReserves(
        [mockReserve],
        mockReserve.lastUpdateTimestamp + 1,
        [mockReserve as any]
      )[0];
      const second = formatReserves(
        [mockReserve],
        mockReserve.lastUpdateTimestamp + 2,
        [mockReserve as any]
      )[0];

      expect(new BigNumber(second.totalDebt).gte(first.totalDebt)).toBe(true);
    });

    it('should compute collateral balance from blockchain data', () => {
      // data exported from user 0xa5a69107816c5e3dfa5561e6b621dfe6294f6e5b
      // at block number: 11581421
      // reserve: YFI
      const scaledATokenBalance = '161316503206059870';
      const liquidityIndex = '1001723339432542553527150680';
      const currentLiquidityRate = '22461916953455574582370088';
      const lastUpdateTimestamp = 1609673617;
      // at a later time, but on the same block
      // expected balance computed with hardhat
      const currentTimestamp = 1609675535;
      const expectedATokenBalance = '161594727054623229';
      const underlyingBalance = getCompoundedBalance(
        scaledATokenBalance,
        liquidityIndex,
        currentLiquidityRate,
        lastUpdateTimestamp,
        currentTimestamp
      ).toString();
      expect(underlyingBalance).toBe(expectedATokenBalance);
    });
  });
});
