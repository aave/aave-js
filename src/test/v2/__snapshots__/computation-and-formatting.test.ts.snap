// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "availableBorrowsETH": "0",
  "currentLiquidationThreshold": "0",
  "currentLoanToValue": "0",
  "healthFactor": "0",
  "id": "0cd96fb5ee9616f64d892644f53f35be4f90xff795577d9ac8bd7d90ee22b6c1703490b6512fd0x88757f2f99175387ab4c6a4b3067c77a695b0349",
  "reservesData": Array [
    Object {
      "principalStableDebt": "0",
      "reserve": Object {
        "decimals": 18,
        "id": "0xff795577d9ac8bd7d90ee22b6c1703490b6512fd0x88757f2f99175387ab4c6a4b3067c77a695b0349",
        "lastUpdateTimestamp": 1607000068,
        "liquidityRate": "0.02707021029693311716",
        "name": "",
        "reserveLiquidationBonus": "0.05",
        "symbol": "DAI",
        "underlyingAsset": "0xff795577d9ac8bd7d90ee22b6c1703490b6512fd",
      },
      "scaledATokenBalance": "0",
      "scaledVariableDebt": "137602026075945229933190",
      "stableBorrowLastUpdateTimestamp": 0,
      "stableBorrowRate": "0",
      "stableBorrows": "0",
      "stableBorrowsETH": "0",
      "stableBorrowsUSD": "0",
      "totalBorrows": "137635.694716770188069538",
      "totalBorrowsETH": "224.903606951938325815",
      "totalBorrowsUSD": "375539.3154302718",
      "underlyingBalance": "0",
      "underlyingBalanceETH": "0",
      "underlyingBalanceUSD": "0",
      "usageAsCollateralEnabledOnUser": false,
      "variableBorrowIndex": "0",
      "variableBorrows": "137635.694716770188069538",
      "variableBorrowsETH": "224.903606951938325815",
      "variableBorrowsUSD": "375539.3154302718",
    },
  ],
  "totalBorrowsETH": "224.903606951938325815",
  "totalBorrowsUSD": "375539.3154302718",
  "totalCollateralETH": "0",
  "totalCollateralUSD": "0",
  "totalLiquidityETH": "0",
  "totalLiquidityUSD": "0",
}
`;

exports[`computations and formattings formatReserves should return plausible results 1`] = `
Object {
  "aTokenAddress": "0xdCf0aF9e59C002FA3AA091a46196b37530FD48a8",
  "availableLiquidity": "43133.641118657852003256",
  "averageStableRate": "109284236984257451326752610",
  "avg30DaysLiquidityRate": undefined,
  "avg30DaysVariableBorrowRate": undefined,
  "baseLTVasCollateral": "0.75",
  "baseVariableBorrowRate": "109284236984257451326752610",
  "borrowingEnabled": true,
  "decimals": 18,
  "id": "0xff795577d9ac8bd7d90ee22b6c1703490b6512fd0x88757f2f99175387ab4c6a4b3067c77a695b0349",
  "isActive": true,
  "isFrozen": false,
  "lastUpdateTimestamp": 1606992400,
  "liquidityIndex": "1.00016444737961059057",
  "liquidityRate": "0.02677620073531209306",
  "name": "",
  "optimalUtilisationRate": "109284236984257451326752610",
  "price": Object {
    "priceInEth": "0.00163405",
  },
  "reserveFactor": "0.1",
  "reserveLiquidationBonus": "0.05",
  "reserveLiquidationThreshold": "0.8",
  "stableBorrowRate": "0.10928437169401419784",
  "stableBorrowRateEnabled": true,
  "stableDebtLastUpdateTimestamp": 1606992400,
  "stableDebtTokenAddress": "0x3B91257Fe5CA63b4114ac41A0d467D25E2F747F3",
  "stableRateSlope1": "20000000000000000000000000",
  "stableRateSlope2": "750000000000000000000000000",
  "symbol": "DAI",
  "totalDebt": "145531.711188584838529666",
  "totalLiquidity": "188665.352307242690532922",
  "totalPrincipalStableDebt": "1",
  "totalScaledVariableDebt": "145496.831599325217573288",
  "totalStableDebt": "1.00000693078484952",
  "totalVariableDebt": "145530.711181654053680146",
  "underlyingAsset": "0xff795577d9ac8bd7d90ee22b6c1703490b6512fd",
  "usageAsCollateralEnabled": true,
  "utilizationRate": "0.77137486776896665556",
  "variableBorrowIndex": "1.00023285443371120965",
  "variableBorrowRate": "0.03856874338802839568",
  "variableDebtTokenAddress": "0xEAbBDBe7aaD7d5A278da40967E62C8c8Fe5fAec8",
  "variableRateSlope1": "40000000000000000000000000",
  "variableRateSlope2": "750000000000000000000000000",
}
`;
